Sming/Makefile-app.mk

```diff
@@ -4,94 +4,329 @@
+APPCODE					+= $$(CMP_$1_APPCODE)
+endif
+DEPENDENCIES			+= $$(CMP_$1_DEPENDS)
+CONFIG_VARS				+= $$(CMP_$1_VARS)
+CMP_$1_INCDIRS			:= $$(call AbsoluteSourcePath,$2,$$(COMPONENT_INCDIRS))
+COMPONENTS_EXTRA_INCDIR	+= $$(CMP_$1_INCDIRS)
+ifneq (,$$(COMPONENT_LIBNAME))
+COMPONENT_LIBPATH		:= $$(call GetComponentLibPath,$$(COMPONENT_LIBNAME))
+COMPONENT_TARGETS		+= $$(COMPONENT_LIBPATH)
+COMPONENTS_AR			+= $$(COMPONENT_LIBPATH)
+endif
+CMP_$1_TARGETS			:= $$(COMPONENT_TARGETS)
+ifneq (,$$(COMPONENT_TARGETS))
+COMPONENTS_TO_BUILD		+= $1
 endif
+endef
 
-# => Serial
-CONFIG_VARS	+= COM_SPEED_SERIAL
-CFLAGS		+= -DCOM_SPEED_SERIAL=$(COM_SPEED_SERIAL)
+
+# Application must the first Component to parse
+$(eval $(call ParseComponent,App,$(CURDIR),$(BUILD_BASE),$(abspath $(APP_LIBDIR))))
+# Don't use application includes for other Components, and build last
+COMPONENTS_EXTRA_INCDIR	:=
+COMPONENTS_TO_BUILD		:=
+
+# Append standard search directories to any defined by the application
+COMPONENT_SEARCH_DIRS	+= $(ARCH_COMPONENTS) $(SMING_HOME)/Components $(SMING_HOME)/Libraries
+
+# Each Component search directory can have an optional 'components.mk' file
+# This is primarily for defining dependencies for Arduino Libraries; many of these
+# are submodules so we can establish dependencies without resorting to elaborate
+# recursive fetch/scan strategies.
+$(foreach d,$(COMPONENT_SEARCH_DIRS),$(eval -include $d/components.mk))
+
+# Pull in all Component definitions
+$(foreach c,$(COMPONENTS),$(eval $(call ParseComponent,$c,$(call FindComponentDir,$c),$(SMING_HOME)/$(BUILD_BASE),$(USER_LIBDIR))))
+
+# Pull in any required definitions we haven't already parsed
+DEPENDENCIES			:= $(filter-out $(COMPONENTS),$(DEPENDENCIES))
+$(foreach c,$(DEPENDENCIES),$(eval $(call ParseComponent,$c,$(call FindComponentDir,$c),$(SMING_HOME)/$(BUILD_BASE),$(USER_LIBDIR))))
+COMPONENTS += $(DEPENDENCIES)
+
+# Always build App last
+COMPONENTS_TO_BUILD		+= App
+
+
+export COMPONENTS_EXTRA_INCDIR
+export APPCODE
+export APP_CFLAGS
+
+# Export all config variables
+$(foreach v,$(CONFIG_VARS),$(eval export $v))
+
+
+##@Building
+
+# Apply patch to a submodule
+# $1 -> patch file with relative path
+define ApplyPatch
+	touch $1; \
+	$(GIT) apply -v $1 --ignore-whitespace --whitespace=nowarn
+endef
+
+# If there's a patch for this submodule, apply it
+# We look for patch in .. and in ../.patches
+# $1 -> submodule path
+# $2 -> name of patch file
+define TryApplyPatch
+	cd $1 && if [ -f ../$2 ]; then \
+		$(call ApplyPatch,../$2); \
+	elif [ -f ../.patches/$2 ]; then \
+		$(call ApplyPatch,../.patches/$2); \
+	fi
+endef
+
+# Update and patch submodule
+# Patch file is either in submodule parent directory itself or subdirectory .patches from there
+.PHONY .NOTPARALLEL: %/.submodule
+%/.submodule:
+	$(info )
+	$(info Fetching submodule '$(dir $@)' ...)
+	$(Q) mkdir -p $(dir $@)
+	$(Q) $(GIT) submodule update --init --force --recursive $(dir $@)
```

Comment:
This will not work when the application code is outside of the sming framework folder. One simple fix would be to replace the command above with:

```
$(Q) (cd $(SMING_HOME) && $(GIT) submodule update --init --force --recursive $(dir $@))
```

Reply:
You have changed the line to something different but it still does not work. 
```
cd /path/to/Sming/Sming/Components/spiffs/spiffs/ && git submodule --quiet update --init --force --recursive .
error: pathspec '.' did not match any file(s) known to git.
```

Please, use the command line that I have suggested.  Also update the travis test to copy one of the samples to a folder which is outside of the Sming worktree. This way we should be able to notice issues with the build mechanism.

Reply:
What I want to do here is support Component repositories which don't live in the Sming repo, hence not using `cd $(SMING_HOME)`. I'm guessing it's the missing () subshell bit which fails, so does this work for you?

```
$(Q) (cd $(dir $@) && $(GIT) submodule --quiet update --init --force --recursive .)
```

Reply:
> What I want to do here is support Component repositories which don't live in the Sming repo, 

Ok, that makes sense. I would love to have that working. But at the moment you are already using $(SMING_HOME) in that code block and the block is SMING_HOME only.
```
	$(info Fetching submodule '$(patsubst $(SMING_HOME)/%,%,$(dir $@))' ...)
	$(Q) mkdir -p $(dir $@)
	$(Q) cd $(dir $@) && $(GIT) submodule --quiet update --init --force --recursive .
```
Which means that line 1 also needs to be updated. 
So for me it seemed like 
> so does this work for you?

No, that did not work either.